jetty-http2/http2-client/src/test/java/org/eclipse/jetty/http2/client/ConcurrentStreamCreationTest.java

```diff
@@ -0,0 +1,105 @@
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.HttpVersion;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.http2.api.Session;
+import org.eclipse.jetty.http2.api.Stream;
+import org.eclipse.jetty.http2.api.server.ServerSessionListener;
+import org.eclipse.jetty.http2.frames.HeadersFrame;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Promise;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ConcurrentStreamCreationTest extends AbstractTest
+{
+    @Test
+    public void testConcurrentStreamCreation() throws Exception
+    {
+        int threads = 64;
+        int runs = 1;
+        int iterations = 1024;
+        int total = threads * runs * iterations;
+        CountDownLatch serverLatch = new CountDownLatch(total);
+        start(new ServerSessionListener.Adapter()
+        {
+            @Override
+            public Stream.Listener onNewStream(Stream stream, HeadersFrame frame)
+            {
+                MetaData.Response response = new MetaData.Response(HttpVersion.HTTP_2, HttpStatus.OK_200, new HttpFields());
+                HeadersFrame responseFrame = new HeadersFrame(stream.getId(), response, null, true);
+                stream.headers(responseFrame, Callback.NOOP);
+                serverLatch.countDown();
+                return null;
+            }
+        }, h2 -> h2.setMaxConcurrentStreams(total));
+
+        Session session = newClient(new Session.Listener.Adapter());
+
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        CountDownLatch clientLatch = new CountDownLatch(total);
+        CountDownLatch clientStreamLatch = new CountDownLatch(runs);
+        Promise<Stream> promise = new Promise.Adapter<Stream>()
+        {
+            @Override
+            public void succeeded(Stream stream)
+            {
+                clientLatch.countDown();
+            }
+        };
+        IntStream.range(0, threads).forEach(i -> new Thread(() ->
+        {
+            try
+            {
+                barrier.await();
+                IntStream.range(0, runs).forEach(j ->
+                        IntStream.range(0, iterations).forEach(k ->
+                        {
+                            MetaData.Request request = newRequest("GET", new HttpFields());
+                            HeadersFrame requestFrame = new HeadersFrame(request, null, true);
+                            session.newStream(requestFrame, promise, new Stream.Listener.Adapter()
+                            {
+                                @Override
+                                public void onClosed(Stream stream)
+                                {
+                                    clientStreamLatch.countDown();
```

Comment:
Ah, no. The "close" event happens before notifying the listener of the response.
So this really needs to be a response latch, and you should override `onHeaders()` and check that you got a response code 200, then countDown the response latch.

Reply:
Ok, done.